File: src/types/activityGroup.ts
import type { ActivityItem } from "./activity";
import type { GitRefFilter } from "./project";

export type ActivityGroupItemInput = {
  sourceType: string;
  sourceId: string;
  occurredAt: string;
  summarySnapshot: string;
};

export type ActivityGroupItem = ActivityGroupItemInput & {
  id: string;
  groupId: string;
  activity?: ActivityItem | null;
  createdAt: string;
};

export type ActivityGroup = {
  id: string;
  projectId?: string | null;
  projectName?: string | null;
  workspaceId?: string | null;
  workspaceName?: string | null;
  projectCount: number;
  projects: Array<{
    projectId: string;
    projectName: string;
  }>;
  title: string;
  summary?: string | null;
  startDate: string;
  endDate: string;
  source: "ai" | "local_rule" | string;
  confidence: number;
  includedInReport: boolean;
  fingerprint?: string | null;
  algorithmVersion?: string | null;
  confidenceLabel: "strong" | "likely" | "needs_review" | string;
  rationaleJson?: string | null;
  reportSummary?: string | null;
  locked: boolean;
  userEditedAt?: string | null;
  reviewStatus: "draft" | "needs_review" | "reviewed" | string;
  titleConfidence?: number | null;
  titleConfidenceLabel?: "strong" | "likely" | "needs_review" | string | null;
  titleQualityLabel?:
    | "report_ready"
    | "acceptable"
    | "technically_correct_but_weak"
    | "needs_user_review"
    | "fallback_only"
    | "rejected"
    | string
    | null;
  titleStrategy?: string | null;
  titleRationaleJson?: string | null;
  titleCandidatesJson?: string | null;
  createdAt: string;
  updatedAt: string;
  items: ActivityGroupItem[];
};

export type ListActivityGroupsInput = {
  from: string;
  to: string;
  projectIds?: string[] | null;
  workspaceIds?: string[] | null;
  classification?: string | null;
  gitRefs?: GitRefFilter[] | null;
  worktreePaths?: string[] | null;
  includeHidden?: boolean | null;
};

export type SuggestActivityGroupsInput = Omit<ListActivityGroupsInput, "includeHidden"> & {
  useAi?: boolean | null;
  useEmbeddings?: boolean | null;
};

export type CreateActivityGroupInput = {
  projectId?: string | null;
  workspaceId?: string | null;
  title: string;
  summary?: string | null;
  startDate: string;
  endDate: string;
  source?: string | null;
  confidence?: number | null;
  includedInReport?: boolean | null;
  fingerprint?: string | null;
  algorithmVersion?: string | null;
  confidenceLabel?: string | null;
  rationaleJson?: string | null;
  reportSummary?: string | null;
  locked?: boolean | null;
  reviewStatus?: string | null;
  titleConfidence?: number | null;
  titleConfidenceLabel?: string | null;
  titleQualityLabel?: string | null;
  titleStrategy?: string | null;
  titleClassificationJson?: string | null;
  titleCandidatesJson?: string | null;
  titleRationaleJson?: string | null;
  titleRejectedTermsJson?: string | null;
  items: ActivityGroupItemInput[];
};

export type UpdateActivityGroupInput = Partial<
  Omit<CreateActivityGroupInput, "projectId" | "workspaceId" | "items">
>;

export type ReplaceActivityGroupItemsInput = {
  items: ActivityGroupItemInput[];
};

export type GroupingDiffSnippet = {
  commitHash: string;
  path: string;
  snippet: string;
};

export type GroupingEvidence = {
  group: ActivityGroup;
  reasons: string[];
  changedPaths: string[];
  diffSnippets: GroupingDiffSnippet[];
};

export type MergeActivityGroupsInput = {
  sourceGroupIds: string[];
  title?: string | null;
};

export type SplitActivityGroupInput = {
  itemIds: string[];
  title?: string | null;
};

export type MoveActivityGroupItemInput = {
  itemId: string;
  targetGroupId: string;
};

export type LockActivityGroupInput = {
  locked: boolean;
};

export type TitleCandidate = {
  id: string;
  title: string;
  summary: string;
  reportSummary: string;
  action: string;
  domains: string[];
  strategy: string;
  score: number;
  qualityLabel: string;
  rationale: string[];
};

export type TitleRationale = {
  selectedTitle: string;
  selectedAction: string;
  selectedDomains: string[];
  namingStrategy: string;
  titleConfidence: number;
  titleConfidenceLabel: string;
  titleQualityLabel: string;
  positiveEvidence: string[];
  rejectedTerms: string[];
  rejectedCandidates: string[];
  warnings: string[];
};

export type PreviewActivityGroupTitleResponse = {
  selectedTitle: string;
  selectedSummary: string;
  selectedReportSummary: string;
  titleConfidence: number;
  titleConfidenceLabel: string;
  titleQualityLabel: string;
  namingStrategy: string;
  candidates: TitleCandidate[];
  rationale: TitleRationale;
};

export type RegenerateActivityGroupTitleInput = {
  groupId: string;
  persist: boolean;
  respectUserEdited: boolean;
};

export type SelectActivityGroupTitleCandidateInput = {
  groupId: string;
  candidateTitle: string;
  candidateId?: string | null;
};

export type RecordActivityGroupTitleFeedbackInput = {
  groupId: string;
  eventType: string;
  previousTitle?: string | null;
  newTitle?: string | null;
  previousSummary?: string | null;
  newSummary?: string | null;
  metadataJson?: string | null;
};

File: src-tauri/src/domain/activity_group.rs
use serde::{Deserialize, Serialize};

use crate::domain::activity::ActivityItem;
use crate::domain::git_metadata::GitRefFilter;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListActivityGroupsInput {
    pub from: String,
    pub to: String,
    pub project_ids: Option<Vec<String>>,
    pub classification: Option<String>,
    pub git_refs: Option<Vec<GitRefFilter>>,
    pub worktree_paths: Option<Vec<String>>,
    pub include_hidden: Option<bool>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SuggestActivityGroupsInput {
    pub from: String,
    pub to: String,
    pub project_ids: Option<Vec<String>>,
    pub classification: Option<String>,
    pub git_refs: Option<Vec<GitRefFilter>>,
    pub worktree_paths: Option<Vec<String>>,
    pub use_ai: Option<bool>,
    pub use_embeddings: Option<bool>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateActivityGroupInput {
    pub project_id: Option<String>,
    pub title: String,
    pub summary: Option<String>,
    pub start_date: String,
    pub end_date: String,
    pub source: Option<String>,
    pub confidence: Option<f64>,
    pub included_in_report: Option<bool>,
    pub fingerprint: Option<String>,
    pub algorithm_version: Option<String>,
    pub confidence_label: Option<String>,
    pub rationale_json: Option<String>,
    pub report_summary: Option<String>,
    pub locked: Option<bool>,
    pub review_status: Option<String>,
    pub title_confidence: Option<f64>,
    pub title_confidence_label: Option<String>,
    pub title_quality_label: Option<String>,
    pub title_strategy: Option<String>,
    pub title_classification_json: Option<String>,
    pub title_candidates_json: Option<String>,
    pub title_rationale_json: Option<String>,
    pub title_rejected_terms_json: Option<String>,
    pub items: Vec<ActivityGroupItemInput>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateActivityGroupInput {
    pub title: Option<String>,
    pub summary: Option<String>,
    pub start_date: Option<String>,
    pub end_date: Option<String>,
    pub source: Option<String>,
    pub confidence: Option<f64>,
    pub included_in_report: Option<bool>,
    pub report_summary: Option<String>,
    pub locked: Option<bool>,
    pub review_status: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReplaceActivityGroupItemsInput {
    pub items: Vec<ActivityGroupItemInput>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreviewActivityGroupSuggestionsInput {
    pub from: String,
    pub to: String,
    pub project_ids: Option<Vec<String>>,
    pub classification: Option<String>,
    pub git_refs: Option<Vec<GitRefFilter>>,
    pub worktree_paths: Option<Vec<String>>,
    pub use_embeddings: Option<bool>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MergeActivityGroupsInput {
    pub source_group_ids: Vec<String>,
    pub title: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SplitActivityGroupInput {
    pub item_ids: Vec<String>,
    pub title: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MoveActivityGroupItemInput {
    pub item_id: String,
    pub target_group_id: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LockActivityGroupInput {
    pub locked: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreviewActivityGroupTitleInput {
    pub group_id: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegenerateActivityGroupTitleInput {
    pub group_id: String,
    pub persist: bool,
    pub respect_user_edited: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SelectActivityGroupTitleCandidateInput {
    pub group_id: String,
    pub candidate_title: String,
    pub candidate_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RecordActivityGroupTitleFeedbackInput {
    pub group_id: String,
    pub event_type: String,
    pub previous_title: Option<String>,
    pub new_title: Option<String>,
    pub previous_summary: Option<String>,
    pub new_summary: Option<String>,
    pub metadata_json: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityGroupItemInput {
    pub source_type: String,
    pub source_id: String,
    pub occurred_at: String,
    pub summary_snapshot: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityGroup {
    pub id: String,
    pub project_id: Option<String>,
    pub project_name: Option<String>,
    pub title: String,
    pub summary: Option<String>,
    pub start_date: String,
    pub end_date: String,
    pub source: String,
    pub confidence: f64,
    pub included_in_report: bool,
    pub fingerprint: Option<String>,
    pub algorithm_version: Option<String>,
    pub confidence_label: String,
    pub rationale_json: Option<String>,
    pub report_summary: Option<String>,
    pub locked: bool,
    pub user_edited_at: Option<String>,
    pub review_status: String,
    pub title_confidence: Option<f64>,
    pub title_confidence_label: Option<String>,
    pub title_quality_label: Option<String>,
    pub title_strategy: Option<String>,
    pub title_rationale_json: Option<String>,
    pub title_candidates_json: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub items: Vec<ActivityGroupItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityGroupNarrative {
    pub group_id: String,
    pub title: String,
    pub summary: Option<String>,
    pub report_summary: Option<String>,
    pub title_confidence: f64,
    pub title_confidence_label: String,
    pub title_quality_label: String,
    pub naming_strategy: String,
    pub classification_json: String,
    pub candidates_json: String,
    pub rationale_json: String,
    pub rejected_terms_json: Option<String>,
    pub algorithm_version: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TitleCandidateDto {
    pub id: String,
    pub title: String,
    pub summary: String,
    pub report_summary: String,
    pub action: String,
    pub domains: Vec<String>,
    pub strategy: String,
    pub score: f64,
    pub quality_label: String,
    pub rationale: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TitleRationaleDto {
    pub selected_title: String,
    pub selected_action: String,
    pub selected_domains: Vec<String>,
    pub naming_strategy: String,
    pub title_confidence: f64,
    pub title_confidence_label: String,
    pub title_quality_label: String,
    pub positive_evidence: Vec<String>,
    pub rejected_terms: Vec<String>,
    pub rejected_candidates: Vec<String>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PreviewActivityGroupTitleResponse {
    pub selected_title: String,
    pub selected_summary: String,
    pub selected_report_summary: String,
    pub title_confidence: f64,
    pub title_confidence_label: String,
    pub title_quality_label: String,
    pub naming_strategy: String,
    pub candidates: Vec<TitleCandidateDto>,
    pub rationale: TitleRationaleDto,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityGroupItem {
    pub id: String,
    pub group_id: String,
    pub source_type: String,
    pub source_id: String,
    pub occurred_at: String,
    pub summary_snapshot: String,
    pub activity: Option<ActivityItem>,
    pub created_at: String,
}

#[derive(Debug, Clone)]
pub struct ActivityGroupTitleMemory {
    pub edited_title: String,
    pub edited_summary: Option<String>,
    pub project_id: Option<String>,
    pub evidence_terms: String,
    pub evidence_terms_json: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupingEvidence {
    pub group: ActivityGroup,
    pub reasons: Vec<String>,
    pub changed_paths: Vec<String>,
    pub diff_snippets: Vec<GroupingDiffSnippet>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupingDiffSnippet {
    pub commit_hash: String,
    pub path: String,
    pub snippet: String,
}
